sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala

```diff
@@ -216,65 +218,61 @@ object ColumnStat extends Logging {
+ * We can add other histograms in the future, which will make estimation logic more complicated.
+ * Because we will have to deal with computation between different types of histograms in some
+ * cases, e.g. for join columns.
+ */
+trait Histogram
 
-    col.dataType match {
-      case dt: IntegralType => fixedLenTypeStruct(dt)
-      case _: DecimalType => fixedLenTypeStruct(col.dataType)
-      case dt @ (DoubleType | FloatType) => fixedLenTypeStruct(dt)
-      case BooleanType => fixedLenTypeStruct(col.dataType)
-      case DateType => fixedLenTypeStruct(col.dataType)
-      case TimestampType => fixedLenTypeStruct(col.dataType)
-      case BinaryType | StringType =>
-        // For string and binary type, we don't store min/max.
-        val nullLit = Literal(null, col.dataType)
-        struct(
-          ndv, nullLit, nullLit, numNulls,
-          // Set avg/max size to default size if all the values are null or there is no value.
-          Coalesce(Seq(Ceil(Average(Length(col))), defaultSize)),
-          Coalesce(Seq(Cast(Max(Length(col)), LongType), defaultSize)))
-      case _ =>
-        throw new AnalysisException("Analyzing column statistics is not supported for column " +
-            s"${col.name} of data type: ${col.dataType}.")
-    }
-  }
+/**
+ * Equi-height histogram represents column value distribution by a sequence of buckets. Each bucket
+ * has a value range and contains approximately the same number of rows.
+ * @param height number of rows in each bucket
+ * @param ehBuckets equi-height histogram buckets
+ */
+case class EquiHeightHistogram(height: Double, ehBuckets: Seq[EquiHeightBucket]) extends Histogram {
 
-  /** Convert a struct for column stats (defined in statExprs) into [[ColumnStat]]. */
-  def rowToColumnStat(row: InternalRow, attr: Attribute): ColumnStat = {
-    ColumnStat(
-      distinctCount = BigInt(row.getLong(0)),
-      // for string/binary min/max, get should return null
-      min = Option(row.get(1, attr.dataType)),
-      max = Option(row.get(2, attr.dataType)),
-      nullCount = BigInt(row.getLong(3)),
-      avgLen = row.getLong(4),
-      maxLen = row.getLong(5)
-    )
+  override def toString: String = {
+    def bucketString(bucket: EquiHeightBucket): String = {
+      val sb = new StringBuilder
+      sb.append("Bucket(")
+      sb.append(bucket.lo)
+      sb.append(", ")
+      sb.append(bucket.hi)
+      sb.append(", ")
+      sb.append(bucket.ndv)
+      sb.append(")")
+      sb.toString()
+    }
+    height + ", " + ehBuckets.map(bucketString).mkString(", ")
   }
-
 }
+
+/**
+ * A bucket in an equi-height histogram. We use double type for lower/higher bound for simplicity.
+ * @param lo lower bound of the value range in this bucket
+ * @param hi higher bound of the value range in this bucket
+ * @param ndv approximate number of distinct values in this bucket
+ */
+case class EquiHeightBucket(lo: Double, hi: Double, ndv: Long)
```

Comment:
The field distinctCount in ColumnStat class represents the number of distinct values in a table.  It has a type BigInt.  Here ndv represents the number of distinct values in a histogram bucket.  It has a type Long.  Should we make them have same type?

Reply:
Using Long is to simplify computation. I think Long is enough for ndv, maybe we can change the BigInt type in `ColumnStat` to Long.